application/data-application.R

```diff
@@ -63,15 +62,103 @@ kokoszka_test <- function(X_fpc, Y_fpc, thre_p = 0.99, thre_q = 0.99) {
                            p = p, q = q, data.name = "Y ~ X"))
   class(result) <- "htest"
   return(result)
+  
+}
 
+# Test for no functional effects by Lee et al. (2020)
+FMDD_test <- function(X_fdata, Y_fdata, int_rule = "trapezoid", B = 1e3) {
+  
+  # Check if X_fdata and Y_fdata are functional data as fdata
+  stopifnot(fda.usc::is.fdata(X_fdata))
+  stopifnot(fda.usc::is.fdata(Y_fdata))
+  
+  # Sample size
+  n <- nrow(X_fdata[["data"]])
+  
+  # Check if sample sizes are matched
+  if (n != nrow(Y_fdata[["data"]])) {
+    stop("X_fdata and Y_fdata samples must have the same number of elements")
+  }
+  
+  # Grid points in which our functional samples are valued
+  s <- X_fdata[["argvals"]]
+  t <- Y_fdata[["argvals"]]
+  
+  # Check if grids are equispaced
+  eps <- sqrt(.Machine[["double.eps"]])
+  equispaced_x <- all(abs(diff(s, differences = 2)) < eps)
+  equispaced_y <- all(abs(diff(t, differences = 2)) < eps)
+  
+  # Building matrices A and B
+  a_ij <- b_ij <- A_mat <- B_mat <- matrix(0, n, n)
+  for (i in 1:n) {
+
+    a_ij[i, ] <- sqrt(apply(t(t(X_fdata[["data"]]) - X_fdata[["data"]][i, ])^2,
+                            1, FUN = "integral1D", s, int_rule, equispaced_x))
+    b_ij[i, ] <- apply(t(t(Y_fdata[["data"]]) - Y_fdata[["data"]][i, ])^2,
+                       1, FUN = "integral1D", t, int_rule, equispaced_y)/2
+  }
+  
+  a_i_dot <- (1 / (n - 2)) * rowSums(a_ij)
+  a_dot_j <- (1 / (n - 2)) * colSums(a_ij)
+  a_dot_dot <- (1 / ((n - 2) * (n - 1))) * sum(sum(a_ij))
+  
+  b_i_dot <- (1 / (n - 2)) * rowSums(b_ij)
+  b_dot_j <- (1 / (n - 2)) * colSums(b_ij)
+  b_dot_dot <- (1 / ((n - 2) * (n - 1))) * sum(sum(b_ij))
+  
+  # A and B matrices: null elements in the diagonal
+  A_mat <- a_ij - matrix(rep(a_i_dot, n), nrow = n) -
+    t(matrix(rep(a_dot_j, n), nrow = n)) + a_dot_dot
+  B_mat <- b_ij - matrix(rep(b_i_dot, n), nrow = n) -
+    t(matrix(rep(b_dot_j, n), nrow = n)) + b_dot_dot
+  diag(B_mat) <- diag(A_mat) <- 0
+  
+  # FMDD_n statistic (sample version)
+  FMDD_n <- (1 / (n * (n - 3))) * sum(sum(A_mat * B_mat))
+  FMDD_orig_stat <- n * FMDD_n
+  
+  # Bootstrapped statistics
+  phi <- (1 + sqrt(5))/2
+  prob <- (phi + 2)/5
+  FMDD_star <- FMDD_boot_stats <- numeric(B)
+  for (b in 1:B) {
+    
+    V <- sample(x = c(1 - phi, phi), prob = c(prob, 1 - prob), size = n,
+                replace = TRUE)
+    
+    FMDD_star[b] <- (1 / (n * (n - 3))) * sum(sum(outer(V, V, "*") *
+                                                    A_mat * B_mat))
+    FMDD_boot_stats[b] <- n * FMDD_star[b]
+    
+  }
+  
+  # p-value
+  p_value <- mean(FMDD_orig_stat <= FMDD_boot_stats)
+  
+  # Output
+  result <-
+    structure(list(statistic = c(statistic = FMDD_orig_stat), p.value = p_value,
+                   boot_statistics = FMDD_boot_stats,
+                   method =
+                     paste0("FMDDD (Functional martingale difference ",
```

Comment:
Maybe just "Functional martingale difference FLMFR significance test" (rather than defining an acronym in the title)?